command/agent/config.go

```diff
@@ -658,137 +663,116 @@ func (c *Config) Merge(b *Config) *Config {
+	if err != nil {
+		return fmt.Errorf("Failed to parse RPC advertise address: %v", err)
+	}
+	c.AdvertiseAddrs.RPC = addr
+
+	addr, err = normalizeAdvertise(c.AdvertiseAddrs.Serf, c.Addresses.Serf, c.Ports.Serf, c.DevMode)
+	if err != nil {
+		return fmt.Errorf("Failed to parse Serf advertise address: %v", err)
+	}
+	c.AdvertiseAddrs.Serf = addr
+
+	return nil
+}
+
+// normalizeBind returns a normalized bind address.
 //
-// Returns true if config is ok and false on errors. Since some normalization
-// issues aren't fatal a logger must be provided to emit warnings.
-func (c *Config) normalize(logger func(string), dev bool) bool {
-	// Set the version info
-	c.Revision = c.Revision
-	c.Version = c.Version
-	c.VersionPrerelease = c.VersionPrerelease
-
-	// Normalize addresses
-	bind := c.BindAddr
-	if err := normalizeBind(&c.Addresses.HTTP, bind, &c.Ports.HTTP); err != nil {
-		logger(err.Error())
-		return false
-	}
-	if err := normalizeBind(&c.Addresses.RPC, bind, &c.Ports.RPC); err != nil {
-		logger(err.Error())
-		return false
-	}
-	if err := normalizeBind(&c.Addresses.Serf, bind, &c.Ports.Serf); err != nil {
-		logger(err.Error())
-		return false
-	}
-
-	if err := normalizeAdvertise(&c.AdvertiseAddrs.HTTP, bind, c.Ports.HTTP, dev); err != nil {
-		logger(err.Error())
-		return false
-	}
-	if err := normalizeAdvertise(&c.AdvertiseAddrs.RPC, bind, c.Ports.RPC, dev); err != nil {
-		logger(err.Error())
-		return false
-	}
-	if err := normalizeAdvertise(&c.AdvertiseAddrs.Serf, bind, c.Ports.Serf, dev); err != nil {
-		logger(err.Error())
-		return false
-	}
-
-	if c.Server.EncryptKey != "" {
-		if _, err := c.Server.EncryptBytes(); err != nil {
-			logger(fmt.Sprintf("Invalid encryption key: %s", err))
-			return false
-		}
-		keyfile := filepath.Join(c.DataDir, serfKeyring)
-		if _, err := os.Stat(keyfile); err == nil {
-			logger("WARNING: keyring exists but -encrypt given, using keyring")
-		}
+// If addr is set it is used, if not the default bind address is used.
+func normalizeBind(addr, bind string) string {
+	if addr == "" {
+		return bind
 	}
-	return true
+	return addr
 }
 
-// Use normalizeAdvertise(&config.AdvertiseAddrs.<Service>, <Port>) to
-// retrieve a default address for advertising if one isn't set.
-func normalizeAdvertise(addr *string, bind string, defport int, dev bool) error {
-	if *addr != "" {
+// normalizeAdvertise returns a normalized advertise address.
+//
+// If addr is set, it is used and the default port is appended if no port is
+// set.
+//
+// If addr is not set and bind is a valid address, the returned string is the
+// bind+port.
+//
+// If addr is not set and bind is not a valid advertise address, the hostname
+// is resolved and returned with the port.
+//
+// Loopback is only considered a valid advertise address in dev mode.
+func normalizeAdvertise(addr string, bind string, defport int, dev bool) (string, error) {
+	if addr != "" {
```

Comment:
[summarizing offline discussion here]

Uses it here but hard fails later in rpc/serf code that resolves advertise addresses.

This matches 0.4.1 behavior, and trying to fix rpc & serf so that they don't resolve advertise addresses is out of scope for 0.5.

Leaving as-is.